controllers/alert_controller.go

```diff
@@ -99,68 +124,32 @@ func (r *AlertReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-func (r *AlertReconciler) validate(ctx context.Context, alert v1beta1.Alert) error {
-	var provider v1beta1.Provider
-	providerName := types.NamespacedName{Namespace: alert.Namespace, Name: alert.Spec.ProviderRef.Name}
-	if err := r.Get(ctx, providerName, &provider); err != nil {
-		return fmt.Errorf("failed to get provider %s, error: %w", providerName.String(), err)
-	}
-
-	if !apimeta.IsStatusConditionTrue(provider.Status.Conditions, meta.ReadyCondition) {
-		return fmt.Errorf("provider %s is not ready", providerName.String())
-	}
+func (r *AlertReconciler) reconcile(ctx context.Context, obj *v1beta1.Alert) (ctrl.Result, error) {
+	// Mark the resource as under reconciliation
+	conditions.MarkReconciling(obj, meta.ProgressingReason, "")
 
-	return nil
-}
-
-func (r *AlertReconciler) recordSuspension(ctx context.Context, alert v1beta1.Alert) {
-	if r.MetricsRecorder == nil {
-		return
+	// validate alert spec and provider
+	if err := r.validate(ctx, obj); err != nil {
+		conditions.MarkFalse(obj, meta.ReadyCondition, meta.FailedReason, err.Error())
+		return ctrl.Result{Requeue: true}, err
 	}
-	log := logr.FromContext(ctx)
 
-	objRef, err := reference.GetReference(r.Scheme, &alert)
-	if err != nil {
-		log.Error(err, "unable to record suspended metric")
-		return
-	}
+	conditions.MarkTrue(obj, meta.ReadyCondition, meta.SucceededReason, v1beta1.InitializedReason)
+	ctrl.LoggerFrom(ctx).Info("Alert initialised")
 
-	if !alert.DeletionTimestamp.IsZero() {
-		r.MetricsRecorder.RecordSuspend(*objRef, false)
-	} else {
-		r.MetricsRecorder.RecordSuspend(*objRef, alert.Spec.Suspend)
-	}
+	return ctrl.Result{}, nil
```

Comment:
That's a good point. My instinct here would be to add a `Watches` on the controller for `ProviderRef`, so that we can reconcile alerts whenever there is a change to `ProviderRef` but that might be overkill.  The other option would be to set a default sync interval but I'm less keen on this; is it something we do elsewhere in the flux controllers?

Reply:
@stefanprodan what is the intent of the validation and the reflection of it on this object?

Reply:
People expect to see validation errors on `flux get` hence this check here. Adding a watcher for the provider is the way to go, we already watch providers in this controller so there is no perf impact of doing so.